### PR TITLE
[New Rule] Potential Shadow Credentials added to AD Object

### LINKED ISSUE #1728: [New Rule] Potential Shadow Credentials added to AD Object
## Description

Identify the modification of the msDS-KeyCredentialLink attribute in a Active Directory Computer or User Object.
Attackers can abuse contol over the object and create a key pair, append to raw public key in the attribute, and obtain
persistent and stealthy access to the target object (can be a user or a computer).

## Example Data

```
{
  "_index": ".ds-logs-system.security-default-2022.01.03-000001",
  "_type": "_doc",
  "_id": "wctdln4B7Kld_xqUfsEA",
  "_version": 1,
  "_score": 1,
  "_source": {
    "agent": {
      "hostname": "DCK",
      "name": "DCK",
      "id": "e9c742aa-7cc9-4ec9-b8ca-fe3e493552a8",
      "ephemeral_id": "87fdf286-ff21-4c1b-a4de-a8b57541d92e",
      "type": "filebeat",
      "version": "7.16.3"
    },
    "winlog": {
      "computer_name": "DCK.kerb.com",
      "process": {
        "pid": 508,
        "thread": {
          "id": 6416
        }
      },
      "keywords": [
        "Audit Success"
      ],
      "logon": {
        "id": "0x5f8fa"
      },
      "channel": "Security",
      "event_data": {
        "AttributeLDAPDisplayName": "msDS-KeyCredentialLink",
        "AttributeValue": "B:828:%%14672:CN=TestA,CN=Users,DC=kerb,DC=com",
        "ObjectClass": "user",
        "AttributeSyntaxOID": "2.5.5.7",
        "DSType": "%%14676",
        "AppCorrelationID": "-",
        "ObjectGUID": "{8131C1D3-BE07-40E5-A695-3E5A63F73140}",
        "SubjectLogonId": "0x5f8fa",
        "SubjectUserName": "Administrator",
        "ObjectDN": "CN=TestA,CN=Users,DC=kerb,DC=com",
        "OpCorrelationID": "{FC404F4A-254E-4944-9CF2-F266C265EE0A}",
        "DSName": "kerb.com",
        "SubjectDomainName": "KERB",
        "OperationType": "%%14674",
        "SubjectUserSid": "S-1-5-21-2581715867-511872244-182302923-500"
      },
      "opcode": "Info",
      "record_id": "11090",
      "task": "Directory Service Changes",
      "event_id": "5136",
      "provider_guid": "{54849625-5478-4994-A5BA-3E3B0328C30D}",
      "activity_id": "{F3C868D2-122A-0000-FB68-C8F32A12D801}",
      "api": "wineventlog",
      "provider_name": "Microsoft-Windows-Security-Auditing"
    },
    "log": {
      "level": "information"
    },
    "elastic_agent": {
      "id": "e9c742aa-7cc9-4ec9-b8ca-fe3e493552a8",
      "version": "7.16.3",
      "snapshot": false
    },
    "message": "A directory service object was modified.\n\t\nSubject:\n\tSecurity ID:\t\tS-1-5-21-2581715867-511872244-182302923-500\n\tAccount Name:\t\tAdministrator\n\tAccount Domain:\t\tKERB\n\tLogon ID:\t\t0x5F8FA\n\nDirectory Service:\n\tName:\tkerb.com\n\tType:\tActive Directory Domain Services\n\t\nObject:\n\tDN:\tCN=TestA,CN=Users,DC=kerb,DC=com\n\tGUID:\t{8131C1D3-BE07-40E5-A695-3E5A63F73140}\n\tClass:\tuser\n\t\nAttribute:\n\tLDAP Display Name:\tmsDS-KeyCredentialLink\n\tSyntax (OID):\t2.5.5.7\n\tValue:\tB:828:<Binary>:CN=TestA,CN=Users,DC=kerb,DC=com\n\t\nOperation:\n\tType:\tValue Added\n\tCorrelation ID:\t{FC404F4A-254E-4944-9CF2-F266C265EE0A}\n\tApplication Correlation ID:\t-",
    "input": {
      "type": "winlog"
    },
    "@timestamp": "2022-01-26T12:30:15.535Z",
    "ecs": {
      "version": "1.12.0"
    },
    "data_stream": {
      "namespace": "default",
      "type": "logs",
      "dataset": "system.security"
    },
    "host": {
      "hostname": "DCK",
      "os": {
        "build": "14393.693",
        "kernel": "10.0.14393.693 (rs1_release.161220-1747)",
        "name": "Windows Server 2016 Datacenter Evaluation",
        "type": "windows",
        "family": "windows",
        "version": "10.0",
        "platform": "windows"
      },
      "ip": [
        "192.168.15.39"
      ],
      "name": "DCK.kerb.com",
      "id": "2d37ee2b-e19c-495f-8540-45bfd43dbb2b",
      "mac": [
        "08:00:27:d3:c9:65"
      ],
      "architecture": "x86_64"
    },
    "event": {
      "agent_id_status": "verified",
      "ingested": "2022-01-26T12:30:16Z",
      "code": "5136",
      "provider": "Microsoft-Windows-Security-Auditing",
      "created": "2022-01-26T12:30:16.577Z",
      "kind": "event",
      "action": "Directory Service Changes",
      "dataset": "system.security",
      "outcome": "success"
    }
  },
  "fields": {
    "elastic_agent.version": [
      "7.16.3"
    ],
    "host.os.name.text": [
      "Windows Server 2016 Datacenter Evaluation"
    ],
    "winlog.provider_guid": [
      "{54849625-5478-4994-A5BA-3E3B0328C30D}"
    ],
    "winlog.provider_name": [
      "Microsoft-Windows-Security-Auditing"
    ],
    "winlog.event_data.DSType": [
      "%%14676"
    ],
    "host.hostname": [
      "DCK"
    ],
    "winlog.computer_name": [
      "DCK.kerb.com"
    ],
    "host.mac": [
      "08:00:27:d3:c9:65",
      "08:00:27:b5:de:43",
      "00:00:00:00:00:00:00:e0",
      "00:00:00:00:00:00:00:e0",
      "00:00:00:00:00:00:00:e0"
    ],
    "winlog.process.pid": [
      508
    ],
    "winlog.event_data.AppCorrelationID": [
      "-"
    ],
    "winlog.keywords": [
      "Audit Success"
    ],
    "host.os.version": [
      "10.0"
    ],
    "winlog.record_id": [
      "11090"
    ],
    "winlog.logon.id": [
      "0x5f8fa"
    ],
    "host.os.name": [
      "Windows Server 2016 Datacenter Evaluation"
    ],
    "log.level": [
      "information"
    ],
    "agent.name": [
      "DCK"
    ],
    "host.name": [
      "DCK.kerb.com"
    ],
    "event.agent_id_status": [
      "verified"
    ],
    "event.kind": [
      "event"
    ],
    "winlog.activity_id": [
      "{F3C868D2-122A-0000-FB68-C8F32A12D801}"
    ],
    "event.outcome": [
      "success"
    ],
    "winlog.event_data.AttributeLDAPDisplayName": [
      "msDS-KeyCredentialLink"
    ],
    "host.os.type": [
      "windows"
    ],
    "input.type": [
      "winlog"
    ],
    "agent.hostname": [
      "DCK"
    ],
    "data_stream.type": [
      "logs"
    ],
    "host.architecture": [
      "x86_64"
    ],
    "winlog.event_data.AttributeSyntaxOID": [
      "2.5.5.7"
    ],
    "winlog.event_data.ObjectGUID": [
      "{8131C1D3-BE07-40E5-A695-3E5A63F73140}"
    ],
    "event.provider": [
      "Microsoft-Windows-Security-Auditing"
    ],
    "event.code": [
      "5136"
    ],
    "agent.id": [
      "e9c742aa-7cc9-4ec9-b8ca-fe3e493552a8"
    ],
    "ecs.version": [
      "1.12.0"
    ],
    "event.created": [
      "2022-01-26T12:30:16.577Z"
    ],
    "winlog.event_data.OperationType": [
      "%%14674"
    ],
    "agent.version": [
      "7.16.3"
    ],
    "host.os.family": [
      "windows"
    ],
    "winlog.event_data.SubjectUserSid": [
      "S-1-5-21-2581715867-511872244-182302923-500"
    ],
    "winlog.process.thread.id": [
      6416
    ],
    "host.os.build": [
      "14393.693"
    ],
    "host.ip": [
      "192.168.15.39"
    ],
    "agent.type": [
      "filebeat"
    ],
    "event.module": [
      "system"
    ],
    "winlog.event_data.SubjectLogonId": [
      "0x5f8fa"
    ],
    "winlog.event_data.DSName": [
      "kerb.com"
    ],
    "host.os.kernel": [
      "10.0.14393.693 (rs1_release.161220-1747)"
    ],
    "winlog.api": [
      "wineventlog"
    ],
    "elastic_agent.snapshot": [
      false
    ],
    "host.id": [
      "2d37ee2b-e19c-495f-8540-45bfd43dbb2b"
    ],
    "winlog.event_data.ObjectClass": [
      "user"
    ],
    "winlog.event_data.OpCorrelationID": [
      "{FC404F4A-254E-4944-9CF2-F266C265EE0A}"
    ],
    "winlog.task": [
      "Directory Service Changes"
    ],
    "elastic_agent.id": [
      "e9c742aa-7cc9-4ec9-b8ca-fe3e493552a8"
    ],
    "winlog.event_data.SubjectUserName": [
      "Administrator"
    ],
    "data_stream.namespace": [
      "default"
    ],
    "winlog.event_data.ObjectDN": [
      "CN=TestA,CN=Users,DC=kerb,DC=com"
    ],
    "message": [
      "A directory service object was modified.\n\t\nSubject:\n\tSecurity ID:\t\tS-1-5-21-2581715867-511872244-182302923-500\n\tAccount Name:\t\tAdministrator\n\tAccount Domain:\t\tKERB\n\tLogon ID:\t\t0x5F8FA\n\nDirectory Service:\n\tName:\tkerb.com\n\tType:\tActive Directory Domain Services\n\t\nObject:\n\tDN:\tCN=TestA,CN=Users,DC=kerb,DC=com\n\tGUID:\t{8131C1D3-BE07-40E5-A695-3E5A63F73140}\n\tClass:\tuser\n\t\nAttribute:\n\tLDAP Display Name:\tmsDS-KeyCredentialLink\n\tSyntax (OID):\t2.5.5.7\n\tValue:\tB:828:<Binary>:CN=TestA,CN=Users,DC=kerb,DC=com\n\t\nOperation:\n\tType:\tValue Added\n\tCorrelation ID:\t{FC404F4A-254E-4944-9CF2-F266C265EE0A}\n\tApplication Correlation ID:\t-"
    ],
    "winlog.event_id": [
      "5136"
    ],
    "event.ingested": [
      "2022-01-26T12:30:16.000Z"
    ],
    "event.action": [
      "Directory Service Changes"
    ],
    "winlog.event_data.AttributeValue": [
      "B:828:%%14672:CN=TestA,CN=Users,DC=kerb,DC=com"
    ],
    "@timestamp": [
      "2022-01-26T12:30:15.535Z"
    ],
    "winlog.channel": [
      "Security"
    ],
    "host.os.platform": [
      "windows"
    ],
    "data_stream.dataset": [
      "system.security"
    ],
    "winlog.opcode": [
      "Info"
    ],
    "agent.ephemeral_id": [
      "87fdf286-ff21-4c1b-a4de-a8b57541d92e"
    ],
    "winlog.event_data.SubjectDomainName": [
      "KERB"
    ],
    "event.dataset": [
      "system.security"
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by w0rk3r**
## Issues

Resolves #1728 

## Summary

Identify the modification of the msDS-KeyCredentialLink attribute in a Active Directory Computer or User Object.
Attackers can abuse contol over the object and create a key pair, append to raw public key in the attribute, and obtain
persistent and stealthy access to the target object (can be a user or a computer).

### CODE REVIEW

File: rules/windows/credential_access_shadow_credentials.toml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+creation_date = "2022/01/26"
+maturity = "production"
+updated_date = "2022/01/26"
+
+[rule]
+author = ["Elastic"]
+description = """
+Identify the modification of the msDS-KeyCredentialLink attribute in a Active Directory Computer or User Object.
+Attackers can abuse contol over the object and create a key pair, append to raw public key in the attribute, and obtain
+persistent and stealthy access to the target object (can be a user or a computer).
```
Comment (brokensound77) [inline rules/windows/credential_access_shadow_credentials.toml:11]:
```suggestion
Identify the modification of the msDS-KeyCredentialLink attribute in an Active Directory Computer or User Object.
Attackers can abuse control over the object and create a key pair, append to raw public key in the attribute, and obtain
persistent and stealthy access to the target user or computer object.
```

File: rules/windows/credential_access_shadow_credentials.toml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+from = "now-9m"
+index = ["winlogbeat-*", "logs-windows.*", "logs-system.*"]
+language = "kuery"
+license = "Elastic License v2"
+name = "Potential Shadow Credentials added to AD Object"
+note = """## Config
+
+The 'Audit Directory Service Changes' logging policy must be configured for (Success, Failure).
+Steps to implement the logging policy with Advanced Audit Configuration:
+
```
Comment (brokensound77) [inline rules/windows/credential_access_shadow_credentials.toml:28]:
```suggestion
```

File: rules/windows/credential_access_shadow_credentials.toml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+]
+risk_score = 73
+rule_id = "f494c678-3c33-43aa-b169-bb3d5198c41d"
+severity = "high"
+tags = ["Elastic", "Host", "Windows", "Threat Detection", "Credential Access"]
+timestamp_override = "event.ingested"
+type = "query"
+
+query = '''
+event.code:"5136" and winlog.event_data.AttributeLDAPDisplayName : "msDS-KeyCredentialLink"
```
Comment (brokensound77) [inline rules/windows/credential_access_shadow_credentials.toml:60]: nit:
```suggestion
event.code:"5136" and winlog.event_data.AttributeLDAPDisplayName:"msDS-KeyCredentialLink"
```

**Review by brokensound77 — APPROVED** [review]
LGTM after comments 👍

**Review by DefSecSentinel — APPROVED** [review]
LGTM

**Pull request merged**